crates/astria-conductor/src/data_availability.rs

```diff
@@ -0,0 +1,382 @@
+                get_rollups.spawn(
+                    get_rollup(client.clone(), height, data, namespace, assembly_tx.clone())
+                        .in_current_span(),
+                );
+            }
+        }
+    }
+
+    // ensure that the last sender goes out of scope so that block assembler's exit condition fires
+    std::mem::drop(assembly_tx);
+
+    block_assembler
+        .await
+        .wrap_err("failed assembling sequencer block subsets")
+}
+
+async fn get_rollup(
+    client: CelestiaClient,
+    height: u64,
+    data: SignedNamespaceData<SequencerNamespaceData>,
+    namespace: Namespace,
+    block_tx: mpsc::Sender<SequencerBlockSubset>,
+) {
+    let rollup_data = match client.get_rollup_data(height, &data, namespace).await {
+        Err(e) => {
+            warn!(error.message = %e, error.cause = ?e, "failed to get rollup data from celestia");
+            return;
+        }
+        Ok(None) => {
+            info!("could not find rollup data");
+            return;
+        }
+        Ok(Some(rollup_data)) => rollup_data,
+    };
+
+    if block_verifier::validate_rollup_data(&rollup_data, data.data.action_tree_root).is_err() {
+        warn!(
+            "could not validate rollup data given the action tree root of the sequencer data; \
+             dropping the block"
+        )
+    } else {
+        let subset = SequencerBlockSubset {
+            block_hash: data.data.block_hash,
+            header: data.data.header,
+            rollup_transactions: rollup_data.rollup_txs,
+        };
+        if block_tx.send(subset).await.is_err() {
+            warn!("failed sending validated rollup data to block assembler; receiver dropped");
+        }
+    }
+}
+
+async fn assemble_blocks(
+    mut assembly_rx: mpsc::Receiver<SequencerBlockSubset>,
+) -> Vec<SequencerBlockSubset> {
+    let mut blocks = Vec::new();
+    while let Some(subset) = assembly_rx.recv().await {
+        blocks.push(subset)
+    }
+    blocks.sort_unstable_by(|a, b| a.header.height.cmp(&b.header.height));
+    blocks
+}
+
+fn verify_all_datas(
+    datas: Vec<SignedNamespaceData<SequencerNamespaceData>>,
+    block_verifier: BlockVerifier,
+) -> JoinMap<tendermint::Hash, eyre::Result<SignedNamespaceData<SequencerNamespaceData>>> {
+    let mut verification_tasks = JoinMap::new();
+    for data in datas {
+        let block_hash = data.data.block_hash;
+        if verification_tasks.contains_key(&block_hash) {
+            warn!(%block_hash,
+                "more than one sequencer data with the same block hash retrieved from celestia; \
+                 only keeping the first"
+            );
+        } else {
+            let verifier = block_verifier.clone();
+            verification_tasks.spawn(
+                block_hash,
+                async move {
+                    verifier
+                        .validate_signed_namespace_data(&data)
+                        .await
+                        .map(|()| data)
```

Comment:
why the `map`?

Reply:
Because I want the validation to return `Ok(data)` or `Err(error)` for further processing. Since the return value of the validation step is just `Result<(), E>` this moves the data into the right place. I suppose I can use `?` and `Ok(data)` to make that clearer